crates/astria-sequencer/src/fees/mod.rs

```diff
@@ -108,381 +79,379 @@ impl Fee {
 }
 
 #[async_trait::async_trait]
-impl FeeHandler for Transfer {
-    type FeeComponents = TransferFeeComponents;
+pub(crate) trait FeeHandler: Send {
+    /// The Pascal-case type name, e.g. `RollupDataSubmission`.
+    // NOTE: We only require this function due to `IbcRelay` not implementing `Protobuf`.
+    fn name() -> &'static str;
+
+    /// The full name including the protobuf package, e.g.
+    /// `astria.protocol.transaction.v1.RollupDataSubmission`.
+    // NOTE: We only require this function due to `IbcRelay` not implementing `Protobuf`.
+    fn full_name() -> String;
+
+    /// The snake-case type name, e.g. `rollup_data_submission`.
+    fn snake_case_name() -> &'static str;
+
+    fn variable_component(&self) -> u128;
 
-    #[instrument(skip_all, err)]
-    async fn check_and_pay_fees<S: StateWrite>(&self, state: S) -> eyre::Result<()> {
+    fn fee_asset(&self) -> Option<&asset::Denom>;
+
+    #[instrument(skip_all, err(level = Level::WARN))]
+    async fn check_and_pay_fees<'a, S>(&self, mut state: S) -> eyre::Result<()>
+    where
+        S: StateWrite,
+        FeeComponents<Self>: TryFrom<StoredValue<'a>, Error = ErrReport>,
+    {
         let fees = state
-            .get_transfer_fees()
+            .get_fees::<Self>()
+            .await
+            .wrap_err_with(|| format!("error fetching {} fees", Self::name()))?
+            .ok_or_else(|| {
+                eyre!(
+                    "{} fees not found, so this action is disabled",
+                    Self::name()
+                )
+            })?;
+        let Some(fee_asset) = self.fee_asset() else {
+            // If the action has no associated fee asset, there are no fees to pay.
+            return Ok(());
+        };
+
+        let variable_fee = self.variable_component().saturating_mul(fees.multiplier());
+        let total_fees = fees.base().saturating_add(variable_fee);
+        let transaction_context = state
+            .get_transaction_context()
+            .expect("transaction source must be present in state when executing an action");
+        let from = transaction_context.address_bytes();
+        let source_action_index = transaction_context.source_action_index;
+
+        ensure!(
+            state
+                .is_allowed_fee_asset(fee_asset)
+                .await
+                .wrap_err("failed to check allowed fee assets in state")?,
+            "invalid fee asset",
+        );
```

Comment:
I stuck with the logic as currently implemented in `main`, but it does seem like it would be better to do the check as early as possible.  I'll change this as suggested.